File: backend/app/schemas/passport.py
"""Passport / publish / public page shapes."""

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.core.enums import (
    DocumentType,
    ImageType,
    PassportStatus,
    ProductCategory,
    ProductStatus,
    VerificationStatus,
)


class PassportSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    public_uuid: UUID
    version: int
    status: PassportStatus
    verification_status: VerificationStatus
    public_url: str
    qr_code_url: str
    created_at: datetime


class PublishResponse(BaseModel):
    product_id: UUID
    status: ProductStatus
    passport: PassportSummary


class PublicMaterial(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    percentage: Decimal
    country_of_origin: str
    recyclable: bool


class PublicSustainability(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    carbon_footprint: str
    water_consumption: str
    recycled_material_percent: Decimal
    repairability_score: Decimal
    recyclable: bool


class PublicCertification(BaseModel):
    name: str
    issuing_authority: str
    issue_date: date
    expiration_date: date | None
    pdf_url: str


class PublicDocument(BaseModel):
    doc_type: DocumentType
    original_filename: str
    file_url: str


class PublicImage(BaseModel):
    id: UUID
    image_type: ImageType
    sort_order: int
    file_url: str


class PublicProduct(BaseModel):
    name: str
    sku: str
    serial_number: str
    category: ProductCategory
    description: str
    production_date: date
    country_of_origin: str


class PublicPassportResponse(BaseModel):
    public_uuid: UUID
    version: int
    status: PassportStatus
    verification_status: VerificationStatus
    created_at: datetime
    product: PublicProduct
    materials: list[PublicMaterial]
    sustainability: PublicSustainability | None
    certifications: list[PublicCertification]
    documents: list[PublicDocument]
    images: list[PublicImage]

File: frontend/types/passport.ts
import type {
  DocumentType,
  ImageType,
  ProductCategory,
  ProductStatus,
} from "./products";

export type PassportStatus = "active" | "revoked";
export type VerificationStatus = "verified" | "unverified";

export interface PassportSummary {
  id: string;
  public_uuid: string;
  version: number;
  status: PassportStatus;
  verification_status: VerificationStatus;
  public_url: string;
  qr_code_url: string;
  created_at: string;
}

export interface PublishResponse {
  product_id: string;
  status: ProductStatus;
  passport: PassportSummary;
}

export interface PassportVersionItem {
  id: string;
  public_uuid: string;
  version: number;
  status: PassportStatus;
  verification_status: VerificationStatus;
  created_at: string;
}

/** Auth-gated QR PNG plus uuid parsed from Content-Disposition. */
export interface ProductQrPayload {
  blob: Blob;
  publicUuid: string;
  filename: string;
}

export interface PublicMaterial {
  name: string;
  percentage: string | number;
  country_of_origin: string;
  recyclable: boolean;
}

export interface PublicSustainability {
  carbon_footprint: string;
  water_consumption: string;
  recycled_material_percent: string | number;
  repairability_score: string | number;
  recyclable: boolean;
}

export interface PublicCertification {
  name: string;
  issuing_authority: string;
  issue_date: string;
  expiration_date: string | null;
  pdf_url: string;
}

export interface PublicDocument {
  doc_type: DocumentType;
  original_filename: string;
  file_url: string;
}

export interface PublicImage {
  id: string;
  image_type: ImageType;
  sort_order: number;
  file_url: string;
}

export interface PublicProduct {
  name: string;
  sku: string;
  serial_number: string;
  category: ProductCategory;
  description: string;
  production_date: string;
  country_of_origin: string;
}

export interface PublicPassport {
  public_uuid: string;
  version: number;
  status: PassportStatus;
  verification_status: VerificationStatus;
  created_at: string;
  product: PublicProduct;
  materials: PublicMaterial[];
  sustainability: PublicSustainability | null;
  certifications: PublicCertification[];
  documents: PublicDocument[];
  images: PublicImage[];
}
